File: koway/assets/hashroutes.dart
import 'dart:io';
import 'dart:convert';

Future<Map<String, dynamic>> readJson(String filepath) async{
  var input = await File(filepath).readAsString();
  Map<String, dynamic> jsonData = jsonDecode(input);
  return jsonData;
}

Future<Map<String, List<String>>> hashRoutes(Map<String,dynamic> data) async {
  Map<String, List<String>> stopToRoutesMap = {};
  List<dynamic> routesData = data['routes'];

  for (var route in routesData){
    String routeNumber = route['routeNumber'];
    List<dynamic> stops = route['stops'];

    for (var stop in stops){
      String stopName = stop['stop_name'];
      if (stopToRoutesMap.containsKey(stopName)){
        stopToRoutesMap[stopName]!.add(routeNumber);
      } else {
        stopToRoutesMap[stopName] = [routeNumber];
      }
    }
  }
  return stopToRoutesMap;
}

Future<void> writeFormattedJson(Map<String, dynamic> data, String filePath) async {
  var jsonString = JsonEncoder.withIndent(' ').convert(data);
  await File(filePath).writeAsString(jsonString);
  print('Data written to $filePath');

}

void main() async {

  var pathToFile = 'routes.json';
  Map<String, dynamic> data = await readJson(pathToFile);
  Map<String, List<String>> routeData = await hashRoutes(data);
  

  var outputFilePath = 'hashed_routes.json';
  await writeFormattedJson(routeData, outputFilePath);

}

File: scripts/hash_routes.py
import json
import os
from typing import Dict, List, Any

def read_json(filepath: str) -> Dict[str, Any]:
    if not os.path.exists(filepath):
        print(f"Error: File '{filepath}' not found.")
        return {}
    
    with open(filepath, 'r', encoding='utf-8') as file:
        return json.load(file)

def hash_routes(data: Dict[str, Any]) -> Dict[str, List[str]]:
    stop_to_routes_map: Dict[str, List[str]] = {}
    routes_data = data.get('routes', [])

    for route in routes_data:
        route_number = route.get('routeNumber')
        stops = route.get('stops', [])

        for stop in stops:
            stop_name = stop.get('stop_name')
            
            if stop_name in stop_to_routes_map:
                stop_to_routes_map[stop_name].append(route_number)
            else:
                stop_to_routes_map[stop_name] = [route_number]
                
    return stop_to_routes_map

def write_formatted_json(data: Dict[str, Any], file_path: str) -> None:
    with open(file_path, 'w', encoding='utf-8') as file:
        json.dump(data, file, indent=1) 
    print(f'Data written to {file_path}')


def main():
    path_to_file = '../koway/assets/routes.json'
    
    data = read_json(path_to_file)
    
    if not data:
        return

    route_data = hash_routes(data)

    output_file_path = '../koway/assets/hashed_routes.json'
    write_formatted_json(route_data, output_file_path)

if __name__ == "__main__":
    main()